packages/langium/src/lsp/completion/follow-element-computation.ts

```diff
@@ -103,25 +108,88 @@ export function findFirstFeatures(feature: ast.AbstractElement | undefined, card
  * @returns A new feature that could be now optional (`?` or `*`).
  */
 function modifyCardinality(feature: ast.AbstractElement, cardinality: Cardinality, cardinalities: Map<ast.AbstractElement, Cardinality>): ast.AbstractElement {
-    if (isOptional(cardinality)) {
-        if (isArray(feature.cardinality)) {
-            cardinalities.set(feature, '*');
-        } else {
-            cardinalities.set(feature, '?');
-        }
-    }
+    cardinalities.set(feature, cardinality);
     return feature;
 }
 
-function findNextFeaturesInGroup(group: ast.Group, index: number, cardinalities: Map<ast.AbstractElement, Cardinality>): ast.AbstractElement[] {
+function findNextFeaturesInGroup(group: ast.Group, index: number, cardinalities: Map<ast.AbstractElement, Cardinality>, visited: Set<ast.AbstractElement>): ast.AbstractElement[] {
     const features: ast.AbstractElement[] = [];
     let firstFeature: ast.AbstractElement;
     do {
         firstFeature = group.elements[index++];
-        features.push(...findFirstFeatures(firstFeature, cardinalities));
+        if (ast.isAction(firstFeature)) {
+            continue;
+        }
+        features.push(...findFirstFeatures(firstFeature, cardinalities, visited));
         if (!isOptional(firstFeature?.cardinality ?? cardinalities.get(firstFeature))) {
             break;
         }
     } while (firstFeature);
     return features;
 }
+
+interface InterpretationContext {
+    tokens: IToken[]
+    stacks: ast.AbstractElement[][]
+}
+
+function interpreteTokens(context: InterpretationContext): void {
+    while (context.tokens.length > 0) {
+        const token = context.tokens.pop()!;
+        const nextFeatureStacks = findNextFeatureStacks(context.stacks, token);
+        context.stacks = nextFeatureStacks;
+    }
```

Comment:
Is this while loop creating some effect by calling `findNextFeatureStacks`? I'm not quite sure if there's a throwable check or something in there, but if not you could just evaluate the last token instead of the loop.

Reply:
Since we always reassign `context.stacks`, there is a side effect for each iteration of the loop. So we definitely have to look at all tokens in the list. You've actually made me aware of an issue in this implementation, since it just should be a normal `for-each` loop instead of using `pop`.